### DISCUSSION TITLE
Need assistance with debugging the entire POCL project

### DISCUSSION THREAD

**Discussion by dinyy** [pocl/pocl#1351] [Q&A]
As a developer, how to debug the entire POCL project during the development process?

**Comment by dinyy** [comment]
For instance, when I use GDB to debug example0, I cannot step into clBuildProgram. How can I understand the underlying implementation of this interface?
![)IW{}PK1XC$U{EIO1}O{((N](https://github.com/pocl/pocl/assets/107465691/44a368ab-e66f-4fb0-bc8c-e08844da2624)

**Comment by jansol** [comment]
If you build PoCL in a Debug configuration it should Just Work. If you are using the ICD loader, that one is likely built without debug symbols so you will have a level or two of ???????? in your call stacks. I don't think that should stop GDB from stepping into the function (but vscode might not recognize the symbol as a function depending on how the headers / ICD loader define it and not offer you the "step into" action as a result). You could try one of these
* build PoCL without ICD support (if you don't need e.g. the proxy driver)
* tell GDB to create a breakpoint in file:line location and if it asks whether this location is in a library that will be loaded at runtime answer yes. This is what I normally do since it lets me run the program all the way to the location I am actually interested in. If you don't know where that is yet, the file in `lib/CL` with the name that matches the function you are trying to debug is usually a good starting point.

**Comment by dinyy** [comment]
> If you build PoCL in a Debug configuration it should Just Work. If you are using the ICD loader, that one is likely built without debug symbols so you will have a level or two of ???????? in your call stacks. I don't think that should stop GDB from stepping into the function (but vscode might not recognize the symbol as a function depending on how the headers / ICD loader define it and not offer you the "step into" action as a result). You could try one of these
> 
> * build PoCL without ICD support (if you don't need e.g. the proxy driver)
> * tell GDB to create a breakpoint in file:line location and if it asks whether this location is in a library that will be loaded at runtime answer yes. This is what I normally do since it lets me run the program all the way to the location I am actually interested in. If you don't know where that is yet, the file in `lib/CL` with the name that matches the function you are trying to debug is usually a good starting point.

thanks for your help, I will have a try.
